### DISCUSSION TITLE
Targets output versioning

### DISCUSSION THREAD

**Discussion by EKtheSage** [ropensci/targets#411] [General]
Hi,

Thank you for creating this superb package. It has helped me save a ton of time by not rerunning all of the steps in my process. I do have one question about targets output, however, that I am hoping you can help me solve.

My process runs every month, and in the end of the pipeline, it will create a model output. How do I keep all of the model output every month? My understanding of targets is that it will re-run the pipeline and overwrite the existing targets, and my use case is that sometimes I may need to read in the output from prior months for further analysis, for example, calculate the error of the prediction from last month. etc.

**Comment by liutiming** [comment]
If one just needs to version the output, using `tar_map()`  is probably sufficient. But because of the need to use output from previous month for further analysis, this sounds more like a use case of target factory to me.

**Reply by EKtheSage** [reply]
@liutiming could you explain in more detail what a target factory is?

**Comment by wlandau** [comment]
> How do I keep all of the model output every month?

`targets` itself does not support history or data versioning, but there are workarounds. The easiest is probably to create a target that writes a uniquely-named data file or report as an untracked side-effect. Alternatively, like @liutiming suggested, a `tar_map()` could generate one target for each month in a range of months, and that range could steadily increase as the months go by. Related: the `targets` data store works pretty nicely with Git and other VCS when the files are small enough, particularly with something like Git LFS. Example for small data: https://github.com/wlandau/targets-minimal#continuous-deployment

**Reply by EKtheSage** [reply]
My monthly process doesn't use multiple versions of the data. It sounds like what you're suggesting with `tar_map()` is that it can use January and February data in February, and January:March data in March. But my process always uses the latest version of the data, so in January I use data as of January, in February I use data as of February. Would `tar_map()` still work in my case?

**Reply by wlandau** [reply]
Sounds like the `tar_map()` workaround wouldn't work if you no longer have access to past versions of the data you are pulling.

**Comment by EKtheSage** [comment]
@wlandau 

I want to revisit this question.

Now I re-read my original post and I realized I might not have been clear with the question, so let me try again.

I think what I am trying to get at is how to keep the outputs of each run in separate instances? An example is like this link: https://docs.wandb.ai/guides/artifacts/model-versioning

In my use case, I usually will run the same pipeline multiple times in a month, I may change a parameter here and there or the data I use in my pipeline updates daily so running on different days would change the results slightly. The challenge I now face is I want to keep track of all these different runs, to see how a parameter change affects the final result. But with my current setup, I cannot do that as outputs get overwritten with new runs. 

Would the continuous deployment example you showed above work when my output files are 50mb+?

**Reply by wlandau** [reply]
I had a use case like this come up recently, and the [`pins`](https://github.com/rstudio/pins) package worked well. I had a target that produced an R Markdown report and then uploaded it as a pin to a GitHub repo.

```r
# _targets.R
library(targets)
library(tarchetypes)
source("R/functions.R")
list(
  tar_target(data, process_data()),
  tar_render(report, "report.Rmd"),
  tar_target(pin, pin_report(report))
)
```

The `pin` target only runs when there are changes to the `pin_report()` function or the `report` target. Because I used GitHub as the `pins` backend, I could go to the repository and look at the history of the rendered report. Git LFS would allow you to extend a similar technique to 50MB+, either that or choose the AWS S3 backend for [`pins`](https://github.com/rstudio/pins) and enable data versioning in your bucket.
